dask/sizeof.py

```diff
@@ -135,44 +135,57 @@ def register_pandas():
     import numpy as np
     import pandas as pd
 
-    def object_size(x):
-        if not len(x):
+    def object_size(*xs):
+        ncells = sum(len(x) for x in xs)
+        if not ncells:
             return 0
-        sample = np.random.choice(x, size=20, replace=True)
-        sample = list(map(sizeof, sample))
-        return sum(sample) / 20 * len(x)
+
+        # Deduplicate Series of references to the same objects,
+        # e.g. as produced by read_parquet
+        unique_samples = {}
+        for x in xs:
+            sample = np.random.choice(x, size=100, replace=True)
+            for i in sample.tolist():
+                unique_samples[id(i)] = i
+
+        sample_nbytes = sum(sizeof(i) for i in unique_samples.values())
+        return int(sample_nbytes * ncells / (100 * len(xs)))
 
     @sizeof.register(pd.DataFrame)
     def sizeof_pandas_dataframe(df):
-        p = sizeof(df.index)
+        p = sizeof(df.index) + sizeof(df.columns)
+        object_cols = []
         for _, col in df.items():
-            p += col.memory_usage(index=False)
+            p += col.memory_usage(index=False, deep=False)
             if col.dtype == object:
-                p += object_size(col._values)
-        return int(p) + 1000
+                object_cols.append(col._values)
+
+        # Deduplicate repeated views of Series of objects; e.g.
+        # df[["name", "name"]]
+        if object_cols:
+            p += object_size(*object_cols)
+        return p + 128
 
     @sizeof.register(pd.Series)
     def sizeof_pandas_series(s):
-        p = int(s.memory_usage(index=True))
+        p = sizeof(s.index) + s.memory_usage(index=False, deep=False)
```

Comment:
Did not measure MultiIndex correctly.
Did not double-count the overhead of Series+index.